keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/nuts/LeapfrogIntegratorTest.java

```diff
@@ -0,0 +1,104 @@
+package io.improbable.keanu.algorithms.mcmc.nuts;
+
+import com.google.common.collect.ImmutableMap;
+import io.improbable.keanu.KeanuRandom;
+import io.improbable.keanu.algorithms.VariableReference;
+import io.improbable.keanu.network.BayesianNetwork;
+import io.improbable.keanu.network.KeanuProbabilisticModelWithGradient;
+import io.improbable.keanu.tensor.dbl.DoubleTensor;
+import io.improbable.keanu.vertices.dbl.probabilistic.GaussianVertex;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.Collections;
+import java.util.Map;
+
+import static io.improbable.keanu.algorithms.mcmc.nuts.VariableValues.ones;
+import static io.improbable.keanu.algorithms.mcmc.nuts.VariableValues.zeros;
+import static io.improbable.keanu.tensor.dbl.DoubleTensor.scalar;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.greaterThan;
+import static org.junit.Assert.assertEquals;
+
+public class LeapfrogIntegratorTest {
+
+    private GaussianVertex vertex;
+    private KeanuProbabilisticModelWithGradient gradientCalculator;
+    private LeapfrogState start;
+    private LeapfrogIntegrator integrator;
+
+    private double initialPosition = 0.0;
+    private double initialMomentum = 0.5;
+
+    @Before
+    public void setup() {
+        vertex = new GaussianVertex(0, 1);
+
+        gradientCalculator = new KeanuProbabilisticModelWithGradient(
+            new BayesianNetwork(vertex.getConnectedGraph())
+        );
+
+        Map<VariableReference, DoubleTensor> p = ImmutableMap.of(vertex.getId(), scalar(0));
+
+        Potential potential = new AdaptiveQuadraticPotential(zeros(p), ones(p), 1, 100, 101, KeanuRandom.getDefaultRandom());
+
+        start = leapfrogAt(vertex, initialPosition, initialMomentum, potential);
+
+        integrator = new LeapfrogIntegrator(potential);
+    }
+
+    @Test
+    public void canLeapForward() {
+
+        LeapfrogState leap = integrator.step(start, gradientCalculator, 1.0);
+        assertEquals(0.5, leap.getPosition().get(vertex.getId()).scalar(), 1e-6);
+    }
+
+    @Test
+    public void canLeapForwardAndBack() {
+
+        LeapfrogState leap = integrator.step(start, gradientCalculator, 1.0);
+        assertEquals(initialMomentum, leap.getPosition().get(vertex.getId()).scalar(), 1e-6);
+
+        LeapfrogState leapBack = integrator.step(leap, gradientCalculator, -1.0);
+        assertEquals(initialPosition, leapBack.getPosition().get(vertex.getId()).scalar(), 1e-6);
+    }
+
+    @Test
+    public void leapsForwardWithMinimalEnergyLoss() {
+
+        LeapfrogState leap = start;
+        for (int i = 0; i < 1000; i++) {
+            leap = integrator.step(leap, gradientCalculator, 1e-3);
+        }
+
+        assertEquals(start.getEnergy(), leap.getEnergy(), 1e-6);
+    }
+
+    @Test
+    public void doesIncreaseKineticEnergyWhenLogProbIncreases() {
+
+        double startEnergy = start.getKineticEnergy();
```

Comment:
The initial position is 0. The initial momentum is 0.5. When it takes a step with the +0.5 momentum it will go down in logProb as you suspect. But potential energy is -logProb so the potential energy has gone down and therefore the kinetic energy has gone up.

Reply:
`doesDecreaseKineticEnergyWhenLogProbDecreases`

Reply:
Nice catch!